[Enhancement] Mob Types

95CivicSi opened the pull request:
This update gives users the option to block spawning of all mobs or just montsers from selected areas.  It also fixes a problem with coordinate keys in centers.yml. 

## Summary of Changes

- Implemented config.yml
- Created default config.yml
- Added 'All-Mobs' option to config.yml
- Moved 'Disabled-Worlds' to config.yml
- Modified array keys in centers.yml
  - This fixes a problem where adding a center block puts quotation marks around the y coordinate key causing the plugin to not work correctly.
- Updated all methods to use config.yml in conjunction with centers.yml
- ~~Fixed problem where user attempts to use PEXP commands from the console.~~
- Updates to onCreatureSpawn()
  - Added check for event cancelled.
  - Added check for mob type
- Created default behaviors in /pexp add command.
  - If All-Mobs option is not provided in the command, the command will check config.yml for the default option.
  - If config.yml does not have a default option set, the command will default to block all mobs.
- Updates to onInteract()
  - Added check for event cancelled.
  - Updates to be compatible with setCenterBlock() changes.


### Usage
Users upgrading to this portion should be prepared to make changes to their centers.yml file and possibly make changes to the new config.yml file.  Due to the number of changes in centers.yml, it is __REQUIRED__ that users with previous centers.yml files delete previous copies and make new ones based on the new template.  It is possible to update previous centers.yml files by hand.  First, the Disabled-Worlds line needs to be moved to config.yml.  Second, each area that is setup will need the x,y, and z keys changed to xPos, yPos, and zPos respectively.  Lastly, each area will need a key added for 'All-Mobs' with the option set for true or false.  True will mean that no mobs will be allowed to spawn in that area. False means that non-aggress/Animal will be allowed to spawn and monsters will be prohibited.

__Example centers.yml: Before__
```
---
Disabled-Worlds: []

WorldSpawn:
  x: 286
  "y": 71
  z: 290
  level: world
  radius: 100
  
SkyWarsArena1:
  x: -141
  "y": 221
  z: 1537
  level: world
  radius: 150
  
...
```

__Example centers.yml: After__
```
---
WorldSpawn:
  xPos: 286
  yPos: 71
  zPos: 290
  level: world
  radius: 100
  allmobs: true

SkyWarsArena1:
  xPos: -141
  yPos: 221
  zPos: 1537
  level: world
  radius: 150
  allmobs: false
...

```

By default, all new areas created with commands will block all mobs from spawning in those areas.  You can change this by passing an option when using the `/pexp add` command or by changing the value in config.yml.  
Changing the value of `All-Mobs` to `false` in config.yml will change the default option of the add command to allow animals to spawn but not monsters.  
Adding true or false to the end of the add command will set the option specifically for that area. The following command would create a protected area called MyProtectedArea with a radius of 30 blocks and would allow animals to spawn but not monsters.
```/pexp add MyProtectedArea 30 false```

95CivicSi committed:
[Enhancement] Mob Types

This adds new functionality to the plugin making it possible for users to choose whether to block only monsters or all mobs.

This also address an issue when creating centers.yml where the "y" key of a center block would be stored with the quotes causing the plugin to fail in protecting areas.

95CivicSi committed:
Merge pull request #1 from 95CivicSi/master

[Enhancement] Mob Types

95CivicSi committed:
TypeHints

Replaced type hints and other commented information.

95CivicSi committed:
Examples

A little more info for users.

Neerdex commented:
How to download this files???

Neerdex commented:
Some errors in console:
Warning: in_array() expects parameter 2 to be array, boolean given in D:\MCPE\plugins\PEXProtection_v0.1.1\src\RevivalPMMP\PEXProtection\Main.php on line 89

Notice: Undefined index: xPos in D:\MCPE\plugins\PEXProtection_v0.1.1\src\RevivalPMMP\PEXProtection\Main.php on line 94

Notice: Undefined index: yPos in D:\MCPE\plugins\PEXProtection_v0.1.1\src\RevivalPMMP\PEXProtection\Main.php on line 95

Notice: Undefined index: zPos in D:\MCPE\GenisysPRO-TESTING\plugins\PEXProtection_v0.1.1\src\RevivalPMMP\PEXProtection\Main.php on line 96

Notice: Undefined index: level in D:\MCPE\plugins\PEXProtection_v0.1.1\src\RevivalPMMP\PEXProtection\Main.php on line 97

Notice: Undefined index: radius in D:\MCPE\plugins\PEXProtection_v0.1.1\src\RevivalPMMP\PEXProtection\Main.php on line 100

Warning: in_array() expects parameter 2 to be array, boolean given in D:\MCPE\plugins\PEXProtection_v0.1.1\src\RevivalPMMP\PEXProtection\Main.php on line 100

ghost commented:
Work. Thanks.

jasonw4331 commented on src/RevivalPMMP/PEXProtection/Main.php:
```diff
@@ -16,179 +16,206 @@
+    public $centers;
+
+    /** @var array */
+	private $tapping;
```
formatting...

jasonw4331 commented on src/RevivalPMMP/PEXProtection/Main.php:
```diff
@@ -16,179 +16,206 @@
+	private $tapping;
+
+	/** @var string */
+	private $level;
```
formatting...

jasonw4331 commented on src/RevivalPMMP/PEXProtection/Main.php:
```diff
@@ -16,179 +16,206 @@
+                        case "del":
+                        case "remove":
+                        case "rem":
+                        case "clear":
```
Why do these sub commands now require the user to be a player?

95CivicSi replied:
Lack of foresight on my part. Added to the update list.

jasonw4331 commented on src/RevivalPMMP/PEXProtection/Main.php:
```diff
@@ -16,179 +16,206 @@
+     */
+    public function onInteract(PlayerInteractEvent $ev) {
+        $p = $ev->getPlayer();
+        if($this->inTapMode($p)) {
```
what if the event is cancelled?

95CivicSi replied:
Great question.  I didn't consider that. It will be updated shortly.

jasonw4331 reviewed (changes requested):
A few recommendations...

95CivicSi committed:
Updates for PR #15

This update is based on comments from @jasonwynn10. Resourced main.php from master to duplicate formatting and added changes in where necessary.

Reverse changes where the console could not use any PexP commands and restricts only the add command.

Add check for event cancelled in onCreatureSpawn() and onInteract().

95CivicSi committed:
Remove Unneccessary files

Removed auto generated IDE files.

95CivicSi commented:
After going through and looking at what was going on, I'm kind of ashamed that I sent this PR out.  Apparently I was working on an older fork that didn't have the most recent updates from @Sandertv .

The overall comparison looks better now.  Sorry for making this one a hassle.

95CivicSi committed:
Better Explaination of All-Mobs

Updated default comment on All-Mobs in config.yml to better explain what this setting does.

95CivicSi committed:
Update gitignore for .idea files

Added rules for .idea folder in .gitignore

jasonw4331 commented on src/RevivalPMMP/PEXProtection/Main.php:
```diff
@@ -21,16 +21,19 @@ class Main extends PluginBase implements Listener {
-			"Disabled-Worlds" => []
-		]);
+		$this->saveDefaultConfig();
+        $this->centers = new Config($this->getDataFolder() . "centers.yml", Config::YAML);
```
formatting...

jasonw4331 commented on src/RevivalPMMP/PEXProtection/Main.php:
```diff
@@ -47,32 +50,38 @@ public function isCenterBlock(string $blockName): bool {
-		return false;
-	}
+	public function isCenterBlockLocation(Block $location){
+        foreach($this->centers->getAll() as $center){
```
formatting

jasonw4331 commented on src/RevivalPMMP/PEXProtection/Main.php:
```diff
@@ -47,32 +50,38 @@ public function isCenterBlock(string $blockName): bool {
+     * @param bool   $allMobs
+     */
+	public function setCenterBlock(string $blockName, Position $location, int $radius, string $level, bool $allMobs){
+        $this->centers->set($blockName, array(
```
seriously... what's with the formatting issue 🤦‍♂️

jasonw4331 commented on src/RevivalPMMP/PEXProtection/Main.php:
```diff
@@ -124,15 +135,25 @@ public function onCommand(CommandSender $sender, Command $command, string $comma
 							if(isset($args[2])) {
 								$sender->sendMessage(TF::AQUA . "Tap a block to add a protection center block with the name " . $args[1] . "!");
 								$this->tapping[$sender->getName()] = $sender->getName();
+                                $this->level = $sender->getLevel()->getName();
```
okay. seriously? formatting...

jasonw4331 commented on src/RevivalPMMP/PEXProtection/Main.php:
```diff
@@ -124,15 +135,25 @@ public function onCommand(CommandSender $sender, Command $command, string $comma
+                                $this->level = $sender->getLevel()->getName();
 								$this->blockName = $args[1];
 								$this->radius = (int) $args[2];
+                                if(isset($args[3])){
```
wth is this formatting??

jasonw4331 commented on src/RevivalPMMP/PEXProtection/Main.php:
```diff
@@ -175,20 +196,23 @@ public function onCommand(CommandSender $sender, Command $command, string $comma
-		}
-	}
+	public function onInteract(PlayerInteractEvent $ev){
+        if(!$ev->isCancelled()){
```
formatting

jasonw4331 reviewed (changes requested):
Why is the formatting always so terrible??

jasonw4331 commented on resources/config.yml:
```diff
@@ -0,0 +1,9 @@
+
+# Default setting for all-mobs when creating protected areas in centers.yml.
+# True means that all mobs will be blocked from spawning inside the protected area. 
+All-Mobs: true;
```
random semicolon?

95CivicSi replied:
Fixed in https://github.com/95CivicSi/PEXProtection/commit/b36bf392a60e7441e27b39e1ffd3b233f277db74

jasonw4331 committed:
Added settings for mob type specifications
Accounting for more of the PluginCommand's checks
Enabled strict type usage
Fixed a bit of formatting
Config now saves async

jasonw4331 committed:
Added settings for mob type specifications
Accounting for more of the PluginCommand's checks
Enabled strict type usage
Fixed a bit of formatting
Config now saves async

jasonw4331 committed:
Fixed disabled-worlds setting

jasonw4331 committed:
Fixed disabled-worlds setting

95CivicSi committed:
Reformat code for uniform indentation.  Tabs replaced with spaces to match PureEntitiesX.

95CivicSi committed:
Remove semicolon for All-Mobs value

95CivicSi commented:
Indentation completely cleaned in https://github.com/95CivicSi/PEXProtection/commit/f58635e9bec6ccc196123b171ef06e258fa6128f.

Replaced all tabs with spaces to match PureEntitesX indentation.

jasonw4331 committed:
Added more true/false statements in make subcommand
Convert to tabs

jasonw4331 committed:
Added more true/false statements in make subcommand
Convert to tabs

jasonw4331 committed:
Change output from missing second command argument
Use world folder names instead of display names

jasonw4331 committed:
Change output from missing second command argument
Use world folder names instead of display names

jasonw4331 committed:
Merge remote-tracking branch 'origin/Mob-Types' into Mob-Types

95CivicSi committed:
Merge pull request #2 from RevivalPMMP/Mob-Types

Mob types Modifications

95CivicSi commented:
@jasonwynn10 The PR you added on my branch has been merged.  Is there anything else holding this up?

jasonw4331 commented:
All appears good to me.

jasonw4331 reviewed (approved):
Tested?

95CivicSi commented:
Yes. I have it working on my local setup and 2 live servers.
